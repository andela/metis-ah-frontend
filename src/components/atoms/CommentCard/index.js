import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import timePast from 'time_past';
import _ from 'underscore';

const CommentCard = (props) => {
  const { comment } = props;
  const name = `${comment.user.firstname} ${comment.user.lastname}`;
  const userImage = 'https://res.cloudinary.com/dbsxxymfz/image/upload/v1536757459/dummy-profile.png';
  return (
    <Fragment>
      <div id="comments-view">
        <div className="comment-body">
          {_.unescape(comment.content)}
        </div>
        <div className="comment-footer">
          <img src={comment.user.image || userImage} alt="userImage" />
          <p className="userName">
            <Link to={`/users/${comment.user.id}`}>{name.trim() ? name : comment.user.username}</Link>
          </p>
          <p className="time-ago">
            {timePast(comment.createdAt)}
          </p>
        </div>
      </div>
    </Fragment>
  );
};

export default CommentCard;
