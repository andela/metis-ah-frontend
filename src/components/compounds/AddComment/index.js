import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import { addComment } from 'Actions/commentActions';

import InputField from 'Atoms/InputField/InputField';
import { showModal } from 'Actions/authUser';
import Alert from 'Utils/helpers/alert';
import './style.scss';

export class AddComment extends React.Component {
  state = {
    content: ''
  };


  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { content } = this.state;
    const { isAuth, showModal } = this.props;

    if (content.length < 2) {
      return Alert.error('Please, enter a comment');
    }
    if (!isAuth) {
      Alert.info('Hi there, you need to be registered or logged in to comment');
      return showModal();
    }
    const { match, addComment } = this.props;
    const { articleId } = match.params;
    return addComment(content, articleId).then(() => {
      this.setState({
        content: ''
      });
    });
  }

  render() {
    const { content } = this.state;
    const { isCreateCommentLoading } = this.props;
    return (
      <div id="add-comment-view">
        <form onSubmit={this.handleSubmit}>
          <textarea
            onChange={this.handleChange}
            name="content"
            placeholder="Write a comment"
            rows="10"
            cols="30"
            value={content}
          />
          <div className="input-container">
            <i id="spinner" className={isCreateCommentLoading ? 'fas fa-spinner fa-spin' : ''} />
            <InputField
              type="submit"
              value="Comment"
              style={isCreateCommentLoading ? { opacity: 0.3 } : null}
              disabled={isCreateCommentLoading}
            />
          </div>
        </form>
      </div>
    );
  }
}

AddComment.defaultProps = {
  isCreateCommentLoading: false
};

AddComment.propTypes = {
  match: PropTypes.objectOf(PropTypes.any).isRequired,
  addComment: PropTypes.func.isRequired,
  showModal: PropTypes.func.isRequired,
  isAuth: PropTypes.bool.isRequired,
  isCreateCommentLoading: PropTypes.bool
};

const mapStateToProps = state => ({
  isAuth: state.authUser.isAuthenticated,
  isCreateCommentLoading: state.commentReducer.isCreateCommentLoading
});

export default withRouter(connect(
  mapStateToProps,
  { addComment, showModal }
)(AddComment));
