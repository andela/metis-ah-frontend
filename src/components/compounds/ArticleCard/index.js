import React from 'react';
import PropTypes from 'prop-types';
import { dayHelper, monthHelper } from '../../util/date';

const ArticleCard = ({
  article, image, firstname, lastname,
}) => {
  const {
    title, createdAt, description, imageUrl, ratings
  } = article;
  return (
    <div className="column is-one-third">
      <div className="ArticleCard">
        <div
          className="ArticleCard__banner"
          style={{
            backgroundImage: `linear-gradient(to bottom, transparent, #000000),url(${imageUrl})`
          }}
        >
          <div className="ArticleCard__BannerContent">
            <h1 className="ArticleCard__BannerTitle">
              {title}
            </h1>
            <p className="ArticleCard__date">
              {monthHelper(createdAt)}
              {' '}
              {dayHelper(createdAt)}
            </p>
          </div>
          <div className="ArticleCard__overlay is-overlay">
            <div className="ArticleCard__OverlayContent">
              <h1 className="ArticleCard__OverlayTitle">{title}</h1>
              <p className="ArticleCard__OverlayText">
                {description}
              </p>
              <p className="ArticleCard__OverlayDate">
                {monthHelper(createdAt)}
                {' '}
                {dayHelper(createdAt)}
              </p>
            </div>
          </div>
        </div>
        <div className="ArticleCard__footer">
          <div className="columns is-mobile">
            <div className="column is-one-fifth PaddingRight--0 Margin--auto">
              <div className="ArticleCard__FooterThumbnail">
                <img src={image} alt="Article author" />
              </div>
            </div>
            <div className="column is-paddingless Margin--auto is-three-fifths">
              <div className="ArticleCard__FooterName">
                {firstname}
                {' '}
                {lastname}
              </div>
            </div>
            <div className="column Margin--auto is-paddingless">
              <div className="ArticleCard__FooterLikes">
                <i className="fas fa-thumbs-up fa-2x" />
                <span>{ratings || 0}</span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>);
};

ArticleCard.propTypes = {
  article: PropTypes.objectOf(PropTypes.any).isRequired,
  firstname: PropTypes.string.isRequired,
  lastname: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};

export default ArticleCard;
