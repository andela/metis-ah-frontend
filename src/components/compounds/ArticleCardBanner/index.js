import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './style.scss';

const style = image => ({
  background: `linear-gradient(to bottom, transparent, transparent, #000000), url('${image}')`
});

const descriptionStyle = () => ({
  display: 'block',
});

class ArticleCardBanner extends Component {
  state = {
    hover: false,
  }

  clearStyle = () => {
    this.setState({
      hover: true
    });
  };

  returnStyle = () => {
    this.setState({
      hover: false
    });
  };

  render() {
    const { hover } = this.state;
    const { item } = this.props;
    return (
      <div
        className="ArticleCard__banner article-card-banner"
        style={hover ? {} : style(item.banner)}
        onMouseEnter={this.clearStyle}
        onMouseLeave={this.returnStyle}
      >
        <div className="ArticleCard__BannerContent">
          <h3 className="ArticleCard__BannerTitle">
            {item.title}
          </h3>
          <p
            className="ArticleCard__BannerDescription"
            style={!hover ? {} : descriptionStyle()}
          >
            {item.description}
          </p>
          <p className="ArticleCard__date">
            {item.date}
          </p>
        </div>
      </div>
    );
  }
}

ArticleCardBanner.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    likes: PropTypes.number.isRequired,
    banner: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    authorImg: PropTypes.string.isRequired,
  }).isRequired,
};

export default ArticleCardBanner;
