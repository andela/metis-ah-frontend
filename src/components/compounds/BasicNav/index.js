import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

import './style.scss';

const BasicNav = ({
  logo, brandName, classes = 'navbar', id = 'navbar-plain', ...props
}) => (
  <Fragment>
    <nav id={id} {...props} className={classes}>
      <div className="container">
        <Link className="navbar-item" to="/">
          <img id="logo" src={logo} alt="Authors haven logo" />
          <h1 className="brand">{brandName}</h1>
        </Link>
      </div>
    </nav>
  </Fragment>
);

export default BasicNav;
