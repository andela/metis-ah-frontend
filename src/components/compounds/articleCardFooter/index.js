import React from 'react';
import PropTypes from 'prop-types';

import './style';

const computeLikes = (likes) => {
  if (likes >= 1000) {
    let num = likes / 1000;
    num = num.toString();
    num += 'K';

    return num;
  }
};

const articleCardFooter = (props) => {
  const { item } = props;
  return (
    <div className="ArticleCard__footer">
      <div className="ArticleCard__FooterThumbnail">
        <img src={item.authorImg} alt="Article author" />
      </div>
      <div className="ArticleCard__FooterName">
        {item.author}
      </div>
      <div className="ArticleCard__FooterLikes">
        <i className="fas fa-thumbs-up fa-2x" />
        <span>{computeLikes(item.likes)}</span>
      </div>
    </div>
  );
};

articleCardFooter.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    likes: PropTypes.number.isRequired,
    banner: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    authorImg: PropTypes.string.isRequired,
  }).isRequired,
};

export default articleCardFooter;
