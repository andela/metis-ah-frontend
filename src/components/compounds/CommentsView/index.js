import React, { Fragment } from "react";
import { connect } from "react-redux";
import { getComments, clearComment } from "Actions/commentActions";
import { withRouter } from "react-router-dom";

import "./style.scss";
import PropTypes from "prop-types";
import CommentCard from "Atoms/CommentCard";

export class CommentsView extends React.Component {
  state = {};

  componentWillMount() {
    const { clearComment } = this.props;
    return clearComment();
  }

  viewComments = () => {
    const { getComments, match } = this.props;
    const { articleId } = match.params;
    return getComments(articleId);
  };

  render() {
    const { comments, isFetchCommentLoading, loggedInUserId } = this.props;

    return (
      <Fragment>
        <div
          id="comments-view-button"
          onClick={this.viewComments}
          aria-hidden="true"
          style={isFetchCommentLoading ? { opacity: 0.6 } : null}
          disabled={isFetchCommentLoading}
        >
          <h1>
            {
              <i
                id="spinner"
                className={
                  isFetchCommentLoading ? "fas fa-spinner fa-spin" : ""
                }
              />
            }{" "}
            View Comments
          </h1>
        </div>

        {comments.length > 0
          ? comments.map(comment => (
              <CommentCard
                key={comment.id}
                comment={comment}
                articleId={comment.articleId}
                commentId={comment.id}
                loggedInUserId={loggedInUserId}
              />
            ))
          : null}
      </Fragment>
    );
  }
}

CommentsView.defaultProps = {
  comments: [],
  isFetchCommentLoading: false
};

CommentsView.propTypes = {
  getComments: PropTypes.func.isRequired,
  clearComment: PropTypes.func.isRequired,
  match: PropTypes.objectOf(PropTypes.any).isRequired,
  comments: PropTypes.arrayOf(PropTypes.object),
  isFetchCommentLoading: PropTypes.bool
};

const mapStateToProps = state => ({
  comments: state.comments.comments,
  isFetchCommentLoading: state.comments.isFetchCommentLoading,
  loggedInUserId: state.authUser.user.id
});

export default withRouter(
  connect(
    mapStateToProps,
    { getComments, clearComment }
  )(CommentsView)
);
