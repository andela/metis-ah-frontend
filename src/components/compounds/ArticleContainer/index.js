
import React, { Component } from 'react';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ArticleCard from 'Components/atoms/ArticleCard';
import { getArticle } from 'Actions/article';
import sortArrayContent from '../../util/helper';
import './style.scss';

class ArticleContainer extends Component {
  constructor(props) {
    super(props);
    const {
      match: { params: category }
    } = this.props;
    this.state = {
      category
    };
  }

  componentDidMount() {
    const { getArticles, match, categories } = this.props;
    const heroContent = sortArrayContent(categories, match.params.category);
    getArticles(match.params.category, heroContent);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      match: {
        params: { category }
      }
    } = nextProps;
    const { categories } = nextProps;
    const heroContent = sortArrayContent(categories, category);
    if (prevState.category.category !== nextProps.match.params.category) {
      nextProps.getArticles(category, heroContent);
      return {
        category: { category }
      };
    }
    return null;
  }

  render() {
    const { articles, loading } = this.props;

    const mappedArticleCards = articles.map((article) => {
      const {
        id,
        imageUrl,
        slug,
        title,
        description,
        createdAt,
        User: { firstname, lastname }
      } = article;
      return (
        <ArticleCard
          key={slug}
          id={id}
          img={imageUrl}
          title={title}
          description={description}
          name={`${firstname} ${lastname}`}
          time={createdAt}
        />
      );
    });
    return (
      <div className="container article-inner-container">
        {mappedArticleCards}
        {articles.length < 1 && (loading && <div className="article-loader" />)}
        {articles.length < 1
          && (!loading && (
            <div className="article-container-card">
              <i className="fas fa-exclamation notification-icon" />
              <p>
                Oops!, No Article currently exits in this category.
                <br />
                We would like you to be the first to write in this category
              </p>
              <br />
              <Link to="/articles/new">WRITE</Link>
            </div>
          ))}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  loading: state.article.loading,
  articles: state.article.articles,
  categories: state.categories.categories
});

const mapDispatchToProps = dispatch => ({
  getArticles: (categoryId, heroContent) => dispatch(getArticle(categoryId, heroContent))
});

ArticleContainer.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool.isRequired,
  match: PropTypes.objectOf(PropTypes.any).isRequired,
  getArticles: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.object).isRequired
};

ArticleContainer.defaultProps = {
  articles: []
};
export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(ArticleContainer)
);
