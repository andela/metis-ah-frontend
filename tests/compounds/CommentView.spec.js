
import React from 'react';
import { shallow } from 'enzyme';


import { CommentsView } from '../../src/components/compounds/CommentsView';
import CommentCard from '../../src/components/atoms/CommentCard';

const mockFn = jest.fn();


const props = {
  clearComment: mockFn,
  getComments: mockFn,
  isFetchCommentLoading: false,
  comments: [
    {
      id: 9,
      content: 'Thanks',
      edited: false,
      createdAt: '2018-11-13T16:13:29.901Z',
      updatedAt: '2018-11-13T16:13:29.901Z',
      articleId: 170,
      userId: 11,
      user: {
        id: 11,
        username: 'johngorithm',
        firstname: 'JAY',
        lastname: 'OBI',
        image: 'http://res.cloudinary.com/dbsxxymfz/image/upload/v1541429992/es8jfxtycdfjvec6ha1a.png'
      }
    }
  ],
  match: {
    params: {
      articleId: 2
    }
  }
};

const propsWithNoComment = {
  clearComment: mockFn,
  getComments: mockFn,
  isFetchCommentLoading: false,
  comments: [],
  match: {
    params: {
      articleId: 2
    }
  }
};
describe('CommentsView Tests', () => {
  const wrapper = shallow(<CommentsView {...props} />);

  test('renders without crashing', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.instance().props.clearComment).toHaveBeenCalled();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should should spinner when loading', () => {
    wrapper.setProps({ isFetchCommentLoading: true });
    expect(wrapper.instance().props.isFetchCommentLoading).toBe(true);
  });

  test('should close modal when CLOSE button is clicked', () => {
    wrapper.find('#comments-view-button').simulate('click');
    expect(wrapper.instance().props.getComments).toHaveBeenCalledWith(2);
  });

  test('does not render comments when none is exists', () => {
    const wrapper2 = shallow(<CommentsView {...propsWithNoComment} />);
    expect(wrapper2.find(<CommentCard />).exists()).toBe(false);
  });
});
