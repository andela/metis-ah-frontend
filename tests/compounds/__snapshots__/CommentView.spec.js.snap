// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsView Tests should match snapshot 1`] = `
<Fragment>
  <div
    aria-hidden="true"
    disabled={false}
    id="comments-view-button"
    onClick={[Function]}
    style={null}
  >
    <h1>
      <i
        className=""
        id="spinner"
      />
       
      View Comments
    </h1>
  </div>
  <CommentCard
    comment={
      Object {
        "articleId": 170,
        "content": "Thanks",
        "createdAt": "2018-11-13T16:13:29.901Z",
        "edited": false,
        "id": 9,
        "updatedAt": "2018-11-13T16:13:29.901Z",
        "user": Object {
          "firstname": "JAY",
          "id": 11,
          "image": "http://res.cloudinary.com/dbsxxymfz/image/upload/v1541429992/es8jfxtycdfjvec6ha1a.png",
          "lastname": "OBI",
          "username": "johngorithm",
        },
        "userId": 11,
      }
    }
    key="9"
  />
</Fragment>
`;
